Unable to use generics after updating to 0.8 from 0.7

I am attempting to update my codebase to use zerocopy 0.8 for all the wins we get! However, I'm stuck on a build break that I'm having a hard time understanding.

Here is a simplified version of the code causing the issue:

```rust
use zerocopy::{FromBytes, Immutable, KnownLayout, IntoBytes};
use zerocopy_derive::{FromBytes, Immutable, IntoBytes, KnownLayout};

#[derive(IntoBytes, FromBytes, KnownLayout, Immutable)]
#[repr(C, packed)]
struct Header {
    magic: u32,
    size: u32,
}

#[derive(IntoBytes, FromBytes, KnownLayout, Immutable)]
#[repr(C, packed)]
struct Packet<P: PacketPayload> {
    header: Header,
    payload: P
}

trait PacketPayload: IntoBytes + FromBytes + KnownLayout + Immutable {}

#[derive(IntoBytes, FromBytes, KnownLayout, Immutable)]
#[repr(C, packed)]
struct SimplePayload {
    name: [u8; 32]
}

impl PacketPayload for SimplePayload {}


fn main() {
    let packet = Packet {
        header: Header {
            magic: 0,
            size: 0
        },
        payload: SimplePayload {
            name: [0; 32]
        }
    };
    let bytes = packet.as_bytes();
    println!("Bytes: {bytes:?}");
}
```

This is the compilation error I'm seeing:

```
error[E0277]: the size for values of type `<P as KnownLayout>::MaybeUninit` cannot be known at compilation time
  --> src/main.rs:11:32
   |
11 | #[derive(IntoBytes, FromBytes, KnownLayout, Immutable)]
   |                                ^^^^^^^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `Sized` is not implemented for `<P as KnownLayout>::MaybeUninit`
   = note: the last field of a packed struct may only have a dynamically sized type if it does not need drop to be run
   = help: change the field's type to have a statically known size
   = note: this error originates in the derive macro `KnownLayout` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `zerocopy_test` (bin "zerocopy_test") due to 1 previous error
```

The equivalent (in my mind) version of this code compiles in zerocopy 0.7:

```rust
use zerocopy::{FromBytes, FromZeroes, AsBytes};
use zerocopy_derive::{FromBytes, FromZeroes, AsBytes};

#[derive(AsBytes, FromBytes, FromZeroes)]
#[repr(C, packed)]
struct Header {
    magic: u32,
    size: u32,
}

#[derive(AsBytes, FromBytes, FromZeroes)]
#[repr(C, packed)]
struct Packet<P: PacketPayload> {
    header: Header,
    payload: P
}

trait PacketPayload: AsBytes + FromBytes + FromZeroes {}

#[derive(AsBytes, FromBytes, FromZeroes)]
#[repr(C, packed)]
struct SimplePayload {
    name: [u8; 32]
}

impl PacketPayload for SimplePayload {}


fn main() {
    let packet = Packet {
        header: Header {
            magic: 0,
            size: 0
        },
        payload: SimplePayload {
            name: [0; 32]
        }
    };
    let bytes = packet.as_bytes();
    println!("Bytes: {bytes:?}");
}
```

--- a/zerocopy-derive/src/lib.rs
+++ b/zerocopy-derive/src/lib.rs
@@ -332,7 +332,15 @@ fn derive_known_layout_inner(ast: &DeriveInput, _top_level: Trait) -> Result<Tok
                             ::zerocopy::util::macro_util::Field<#leading_field_indices>
                         >::Type
                     >,)*
-                    <#trailing_field_ty as ::zerocopy::KnownLayout>::MaybeUninit
+                    // NOTE(#2302): We wrap in `ManuallyDrop` here in case the
+                    // type we're operating on is both generic and
+                    // `repr(packed)`. In that case, Rust needs to know that the
+                    // type is *either* `Sized` or has a trivial `Drop`.
+                    // `ManuallyDrop` has a trivial `Drop`, and so satisfies
+                    // this requirement.
+                    ::zerocopy::util::macro_util::core_reexport::mem::ManuallyDrop<
+                        <#trailing_field_ty as ::zerocopy::KnownLayout>::MaybeUninit
+                    >
                 )
                 where
                     #trailing_field_ty: ::zerocopy::KnownLayout,

--- a/zerocopy-derive/src/output_tests.rs
+++ b/zerocopy-derive/src/output_tests.rs
@@ -198,12 +198,14 @@ fn test_known_layout() {
                     ::zerocopy::util::macro_util::core_reexport::mem::MaybeUninit<
                         <Foo<T, U> as ::zerocopy::util::macro_util::Field<__Zerocopy_Field_0>>::Type,
                     >,
-                    <<Foo<
-                        T,
-                        U,
-                    > as ::zerocopy::util::macro_util::Field<
-                        __Zerocopy_Field_1,
-                    >>::Type as ::zerocopy::KnownLayout>::MaybeUninit,
+                    ::zerocopy::util::macro_util::core_reexport::mem::ManuallyDrop<
+                        <<Foo<
+                            T,
+                            U,
+                        > as ::zerocopy::util::macro_util::Field<
+                            __Zerocopy_Field_1,
+                        >>::Type as ::zerocopy::KnownLayout>::MaybeUninit
+                    >,
                 )
                 where
                     <Foo<

--- a/zerocopy-derive/tests/struct_known_layout.rs
+++ b/zerocopy-derive/tests/struct_known_layout.rs
@@ -101,3 +101,14 @@ impl WithSelfReference {
 }
 
 util_assert_impl_all!(WithSelfReference: imp::KnownLayout);
+
+// Deriving `KnownLayout` should work with generic `repr(packed)` types. See
+// #2302.
+
+#[derive(imp::KnownLayout)]
+#[repr(C, packed)]
+struct Packet<P> {
+    payload: P,
+}
+
+util_assert_impl_all!(Packet<imp::u8>: imp::KnownLayout);

